t/01-test_needle.t

```diff
@@ -37,6 +40,56 @@ throws_ok(
     'died when constructing needle without prior call to needle::init()'
 );
 
+subtest 'needle JSON file not under needle directory' => sub {
+    my $misc_needles_dir = Cwd::cwd;
+    needle::set_needles_dir($misc_needles_dir);
+    my $invalid_json_path = 'invalid/path/to/file.json';
+    throws_ok {
+        needle->new($invalid_json_path);
+    } qr/Needle $invalid_json_path is not under needle directory $misc_needles_dir/,
+      'throws error when needle JSON file is not under needle directory';
+};
+
+subtest 'handle broken JSON file' => sub {
+    my $sandbox = tempdir(CLEANUP => 1);
+    needle::set_needles_dir($sandbox);
+    my $broken_json_path = path($sandbox, 'broken.json');
+
+    path($broken_json_path)->spew('{ "tags": ["test');
+
+    like(warning {
+            my $needle = needle->new($broken_json_path->basename);
+            is($needle, undef, 'needle object not created with broken JSON');
+    }, qr/broken json.*broken\.json/, 'warning shown for broken JSON file');
+};
+
+subtest 'handle tags duplicates' => sub {
+    my $sandbox = tempdir(CLEANUP => 1);
+    needle::set_needles_dir($sandbox);
+    my $tag_json_path = path($sandbox, 'tag.json');
+    my $filename = path($sandbox, 'tag.png');
+
+    sysopen(my $fh, $filename, O_WRONLY | O_CREAT, 0777) or die "Cannot create file '$filename': $!"; close($fh);
```

Comment:
Why are you using sysopen here and not just a normal open?

Reply:
No real reason, I've just found out that method to create a new file. Do you think something like this could be better?
```suggestion
    open(my $fh, '>', $tag_png_path) or die "Cannot open file '$tag_png_path': $!"; close($fh);
```

Reply:
Or just use
```
$filename->spew('');
```
You are using that for the json file, so why not for the png file as well

Reply:
I thought that works only for json encoding, but the subroutine takes string so make sense to just use also to create file